retrieval/discovery/serverset.go

```diff
@@ -327,12 +329,13 @@ func (tc *zookeeperTreeCache) recursiveNodeUpdate(path string, node *zookeeperTr
 				children: map[string]*zookeeperTreeCacheNode{},
 				done:     make(chan struct{}, 1),
 			}
-		}
-		err = tc.recursiveNodeUpdate(path+"/"+child, node.children[child])
-		if err != nil {
-			return err
+			err = tc.recursiveNodeUpdate(path+"/"+child, node.children[child])
```

Comment:
Zookeeper silently merges events, so the event on the parent node might represent deletion of the whole tree and recreating it. Accordingly we need to always refresh all descendants when we get an update about an ancestor.